daemon/mgr/image.go

```diff
@@ -283,6 +288,84 @@ func (mgr *ImageManager) AddTag(ctx context.Context, sourceImage string, targetT
 	return err
 }
 
+// ImageHistory returns image history by reference.
+func (mgr *ImageManager) ImageHistory(ctx context.Context, idOrRef string) ([]types.HistoryResultItem, error) {
+	img, err := mgr.fetchContainerdImage(ctx, idOrRef)
+	if err != nil {
+		return nil, err
+	}
+
+	desc, err := img.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ociImage, err := containerdImageToOciImage(ctx, img)
+	if err != nil {
+		return nil, err
+	}
+
+	// layers info
+	cs := img.ContentStore()
+	manifest, err := ctrdmetaimages.Manifest(ctx, cs, img.Target(), platforms.Default())
+	if err != nil {
+		return nil, err
+	}
+
+	// diffIDs info
+	diffIDs, err := img.RootFS(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(manifest.Layers) != len(diffIDs) {
+		return nil, errors.New("mismatched image rootfs and manifest layers")
+	}
+
+	ociImageHistory := ociImage.History
+	lenOciImageHistory := len(ociImageHistory)
+	history := make([]types.HistoryResultItem, lenOciImageHistory)
+
+	// Note: ociImage History layers info and manifest layers info are all in order from bottom-most to top-most, but the
+	// user-interactive history is in order from top-most to top-bottom, so we need to reverse ociImage History traverse order.
+	for i := range ociImageHistory {
```

Comment:
could we combine this with the following part which is about to get the size of blobs?